toolchain/semantics/semantics_node_stack.h

```diff
@@ -0,0 +1,145 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEMANTICS_SEMANTICS_NODE_STACK_H_
+#define CARBON_TOOLCHAIN_SEMANTICS_SEMANTICS_NODE_STACK_H_
+
+#include "llvm/ADT/SmallVector.h"
+#include "toolchain/parser/parse_node_kind.h"
+#include "toolchain/parser/parse_tree.h"
+#include "toolchain/semantics/semantics_node.h"
+
+namespace Carbon {
+
+// Wraps the stack of nodes for SemanticsParseTreeHandler.
+//
+// All pushes and pops will be vlogged.
+//
+// Pop APIs will run basic verification:
+//
+// - If receiving a pop_parse_kind, verify that the parse_node being popped is
+//   of pop_parse_kind.
+// - If returning name_id or node_id, verify that they are valid.
+// - If _not_ returning name_id or node_id, verify they are invalid.
```

Comment:
From the naming alone, this distinction was a big surprise.

Would it make sense to use `PopInvalidFor...` when specifically handling the invalid nodes on the stack? Or are those much more common and so it'd be better to have `PopValidFor...`? I mostly feel like the valid vs. invalid distinction should be a bit more visible than whether or not a node_id or name_id is requested...

Reply:
Renamed to PopForSoloParseNode, PopAndDiscardSoloParseNode.

Invalid is used to represent the non-provision of a node_id, rather than the provision of an invalid node_id. I've tried to clarify the comments on node_id to this.